File: python/239_slidingWindowMaximum.py
from collections import deque
class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {integer[]}
    def maxSlidingWindow(self, nums, k):
        res = []
        q = deque()
        
        for i in xrange(0, len(nums)):
            while(q and nums[q[-1]] <= nums[i]):
                q.pop()
            q.append(i)
            if (i >= k -1):
                if (i > k - 1):
                    while (i - k + 1) > q[0]:
                        q.popleft()
                res.append(nums[q[0]])
        
        return res

File: cpp_solutions/239_slidingWindowMaximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        deque<int> q;
        for (int i = 0; i < nums.size(); i++){
            
            while(!q.empty() && nums[q.back()] <= nums[i]){
                q.pop_back();
            }
            q.push_back(i);
            
            if (i >= k - 1){
                if (i > k-1){
                    while(q.front() < i - k + 1){
                        q.pop_front();
                    }
                }
                res.push_back(nums[q.front()]);
            }
        }
        
        return res;
    }
};